examples/baseball.py

```diff
@@ -142,28 +132,20 @@ def run_inference(model, at_bats, hits, rng, args):
     return mcmc.get_samples()
 
 
-# TODO: Consider providing generic utilities for doing predictions
-# and computing posterior log density
 def predict(model, at_bats, hits, z, rng, player_names, train=True):
     header = model.__name__ + (' - TRAIN' if train else ' - TEST')
-    model = handlers.substitute(handlers.seed(model, rng), z)
-    model_trace = handlers.trace(model).get_trace(at_bats)
-    predictions = model_trace['obs']['value']
+    predictions = predictive(rng, model, z, at_bats)['obs']
     print_results('=' * 30 + header + '=' * 30,
                   predictions,
                   player_names,
                   at_bats,
                   hits)
     if not train:
-        model = handlers.substitute(model, z)
-        model_trace = handlers.trace(model).get_trace(at_bats, hits)
-        log_joint = 0.
-        for site in model_trace.values():
-            site_log_prob = site['fn'].log_prob(site['value'])
-            log_joint = log_joint + np.sum(site_log_prob.reshape(site_log_prob.shape[:1] + (-1,)),
-                                           -1)
-        log_post_density = logsumexp(log_joint) - np.log(np.shape(log_joint)[0])
-        print('\nPosterior log density: {:.2f}\n'.format(log_post_density))
+        post_loglik = log_likelihood(model, z, at_bats, hits)['obs']
+        # computes expected log predictive density at each data point
+        exp_log_density = logsumexp(post_loglik, axis=0) - np.log(np.shape(post_loglik)[0])
+        # reports log predictive density of all test points
+        print('\nLog pointwise predictive density: {:.2f}\n'.format(exp_log_density.sum()))
```

Comment:
@neerajprad I changed the log density formula which we need to report. Comparing joint_density of these models is not meaningful because each one has its own latent variables and priors. FYI, log pointwise predictive density can be found at https://arxiv.org/pdf/1307.5928.pdf formula (4).

Reply:
Yes, this sounds correct to me! Thanks for fixing this, I think this was wrong in Pyro and just migrated from there.